packages/material-ui/src/InputLabel/InputLabel.js

```diff
@@ -95,8 +95,13 @@ function InputLabel(props) {
   } = props;
 
   let shrink = shrinkProp;
-  if (typeof shrink === 'undefined' && muiFormControl) {
-    shrink = muiFormControl.filled || muiFormControl.focused || muiFormControl.adornedStart;
+  const isAutoShrink = shrink === 'auto';
+
+  if ((typeof shrink === 'undefined' || isAutoShrink) && muiFormControl) {
+    shrink =
+      muiFormControl.filled ||
+      muiFormControl.focused ||
+      (muiFormControl.adornedStart && !isAutoShrink);
```

Comment:
That is a very confusing path at the moment that is equivalent to:
```js
// assuming `muiFormControl` is some object
if (shrink === undefined) {
  shrink = muiFormControl.filled || muiFormControl.focused || muiFormControl.adornedStart;
} else if (isAutoShrink) {
  shrink = shrink = muiFormControl.filled || muiFormControl.focused;
}
```
or 
```js
if (shrink === undefined) {
  shrink = muiFormControl.filled || muiFormControl.focused;
  if (!isAutoShrink) {
    shrink = shrink || muiFormControl.adornedStart;
  }
}
```
Would you agree that these snippets are easier to reason about?

Reply:
Yes, absolutely. I prefer the 1st one, and also, is double assignment to `shrink` was by mistake?
Shouldn't it be
```
// assuming `muiFormControl` is some object
if (shrink === undefined) {
  shrink = muiFormControl.filled || muiFormControl.focused || muiFormControl.adornedStart;
} else if (isAutoShrink) {
  shrink = muiFormControl.filled || muiFormControl.focused;
}
```

Reply:
My mistake, sorry